third_party/move/move-model/src/builder/model_builder.rs

```diff
@@ -460,6 +457,161 @@ impl<'env> ModelBuilder<'env> {
+        if let Type::Struct(..) = ty {
+            let mut display_ctx = self.env.get_type_display_ctx();
+            let root_entry = self.lookup_struct_entry(root);
+            display_ctx.builder_struct_table = Some(&self.reverse_struct_table);
+            display_ctx.type_param_names = Some(
+                root_entry
+                    .type_params
+                    .iter()
+                    .map(|ty_param| ty_param.0)
+                    .collect_vec(),
+            );
+            ty.display(&display_ctx).to_string()
+        } else {
+            panic!("ICE: calling `get_struct_type_name` with non-struct type")
+        }
+    }
+
+    /// Generates error messages for recursive structs
+    /// `path`: `path[0]` contains `path[1]` ... `path[-1]` contains `this_struct`
+    fn gen_error_msg_for_fields_loop(
+        &self,
+        path: &Vec<(Loc, Symbol, Type)>,
+        this_struct_id: QualifiedId<StructId>,
+        _loc: Loc,
+    ) -> Vec<String> {
+        let mut loop_notes = Vec::new();
+        for i in 0..path.len() - 1 {
+            let parent_name = self.get_struct_type_name(&path[i].2, this_struct_id);
+            let field_name = self.env.symbol_pool().string(path[i].1).to_string();
+            let child_name = self.get_struct_type_name(&path[i + 1].2, this_struct_id);
+            loop_notes.push(format!(
+                "`{}` contains field `{}: {}`...",
+                parent_name, field_name, child_name
+            ));
+        }
+        loop_notes.push(format!(
+            "`{}` contains field `{}: {}`, which forms a loop.",
+            self.get_struct_type_name(&path.last().unwrap().2, this_struct_id),
+            self.env.symbol_pool().string(path.last().unwrap().1),
+            self.get_struct_display_name_simple(this_struct_id)
+        ));
+        loop_notes
+    }
+
+    /// Checks if a field type occurs in its access path,
+    /// and adds diagnostics for recursive struct definitions if `ty` occurs in `path` and `ty` is also the struct type `checking`.
+    ///
+    /// Returns true iff no resursive definition for `checking` is found at this level.
+    ///
+    /// `path`: `path[0]` contains `path[1]` ... `path[-1]` contains `ty`
+    /// `checking`: the root struct which we are checking for recursive definition for
+    /// `loc_checking`: location of the struct we are checking for
+    fn check_recusive_struct_with_parents(
+        &self,
+        ty: &Type,
+        ty_loc: Loc,
+        path: &mut Vec<(Loc, Symbol, Type)>,
+        checking: QualifiedId<StructId>,
+        loc_checking: &Loc,
+    ) -> bool {
+        match ty {
+            Type::Struct(mid, sid, insts) => {
+                let this_struct_id = mid.qualified(*sid);
+                let this_struct_entry = self.lookup_struct_entry(this_struct_id);
+                // checks if `ty` occurs in `path`
+                for (_parent_loc, _field_symbol, parent) in path.iter() {
```

Comment:
nit: consider replacing `parent` with `ancestor` (and similar for other related vars), which is what they are here?

Reply:
done